lib/carrierwave/mongoid.rb

```diff
@@ -166,81 +103,231 @@ def mount_uploaders(column, uploader = nil, options = {}, &block)
+          after_save :"remove_previously_stored_#{column}"
+          after_destroy :"remove_#{column}!"
+
+          class_eval <<-RUBY, __FILE__, (__LINE__ + 1)
+            def #{column}=(new_files)
+              column = _mounter(:#{column}).serialization_column
+
+              previous_uploader_value = read_uploader(column)
+              @_previous_uploader_value_for_#{column} = previous_uploader_value
+
+              write_uploader(column, []) if self.persisted? && read_uploader(column).nil?
+
+              send(:"\#{column}_will_change!")
+
+              super
+            end
+
+            def #{column}_changed?
+              changed_attributes.has_key?("#{column}")
+            end
+
+            def remove_#{column}=(value)
+              if ['1', true].include?(value)
+                column = _mounter(:#{column}).serialization_column
+
+                send(:"\#{column}_will_change!")
+              end
+
+              super
+            end
+
+            # The default Mongoid attribute_will_change! method is not enough
+            # when we want to upload a new file in an existing embedded document.
+            # The custom version of that method forces the callbacks to be
+            # ran and so does the upload.
+            def #{column}_will_change!
+              changed_attributes["#{column}"] = ['_new_']
+            end
+
+            def remove_previously_stored_#{column}
+              before, after = @_previous_changes_for_#{column}
+              # Don't delete if the files had the same name
+              return if before.nil? && after.nil?
+              # Proceed to remove the file, use the original name instead of '_new_'
+              before = @_previous_uploader_value_for_#{column} || before
+              _mounter(:#{column}).remove_previous(Array.wrap(before), Array.wrap(after))
+            end
+
+            def serializable_hash(options = nil)
+              hash = {}
+
+              except = options && options[:except] && Array.wrap(options[:except]).map(&:to_s)
+              only = options && options[:only] && Array.wrap(options[:only]).map(&:to_s)
+
+              self.class.uploaders.each do |column, _uploader|
+                if (!only && !except) || (only && only.include?(column.to_s)) || (except && !except.include?(column.to_s))
+                  next if _mounter(column.to_sym).uploaders.blank?
+                  hash[column.to_s] = _mounter(column.to_sym).uploaders.map(&:serializable_hash)
+                end
+              end
+
+              super(options).merge(hash)
+            end
+
+            def store_previous_changes_for_#{column}
+              @_previous_changes_for_#{column} = changes[_mounter(:#{column}).serialization_column]
+            end
+          RUBY
+        end
+      end
+    end
+
+    if Gem::Version.new(CarrierWave::VERSION) >= Gem::Version.new("3.0.beta")
+      include CarrierWave3
+    else
+      include CarrierWavePre3
     end
```

Comment:
Changes in this file can be better understood by using [?w=1](https://github.com/carrierwaveuploader/carrierwave-mongoid/pull/215/files?w=1).

What I did was introducing new modules `CarrierWave3` and `CarrierWavePre3`, and conditionally include one of them based on the CarrierWave version.